### PR TITLE
Precalculate the dimension and usage hierarchies to simplify lookup paths

### PR DESCRIPTION
**Pull request by mridgway**
```
+---------------------------------------+-----------------+-----------------+
|                                       │ YCB             │ Precalc         |
+---------------------------------------+-----------------+-----------------+
| Base                                  │ 109,473 ops/sec │ 530,714 ops/sec |
+---------------------------------------+-----------------+-----------------+
| y20 (large config / large dimensions) │ 1,370 ops/sec   │ 83,370 ops/sec  |
+---------------------------------------+-----------------+-----------------+
| y20 (large config / small dimensions) │ 25,481 ops/sec  │ 177,123 ops/sec |
+---------------------------------------+-----------------+-----------------+
| y20 (small config / large dimensions) │ 4,133 ops/sec   │ 81,394 ops/sec  |
+---------------------------------------+-----------------+-----------------+
| y20 (small config / small dimensions) │ 25,399 ops/sec  │ 195,728 ops/sec |
+---------------------------------------+-----------------+-----------------+
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -417,6 +419,33 @@ Ycb.prototype = {
...
+        // Assemble map of all dimensions used including ancestry
+        this.dimsUsed = {};
+        for (name in dimsUsed) {
+            if (dimsUsed.hasOwnProperty(name) && this._dimensionHeierarchies.hasOwnProperty(name)) {
+                this.dimsUsed[name] = {};
+                for (i=0; i<dimsUsed[name].length; i += 1) {
+                    value = dimsUsed[name][i];
+                    for (hierarchy in this._dimensionHeierarchies[name]) {
+                        if (this._dimensionHeierarchies[name].hasOwnProperty(hierarchy)) {
+                            if (-1 !== this._dimensionHeierarchies[name][hierarchy].indexOf(value)) {
```
Comment (mridgway) [inline index.js]: This also marks any parents of a referenced dimension to also be marked as used. This allows constant-time lookups in `_makeOrderedLookupList`.

File: index.js [hunk 2]
```diff
@@ -519,39 +548,15 @@ Ycb.prototype = {
...
-            value,
-            used,
-            parts,
-            i,
             chains = {};
 
         for (pos = 0; pos < this.dimensions.length; pos += 1) {
             for (name in this.dimensions[pos]) {
                 if (this.dimensions[pos].hasOwnProperty(name)) {
-                    for (path in this._dimensionPaths[name]) {
```
Comment (mridgway) [inline index.js]: All of this is handled at instantiation time in `_calculateHierarchy`.

**Comment by folta** [comment]
@redonkulus I think you CC'd the wrong person :)

**Comment by redonkulus** [comment]
@folta you are correct, my mistake. @drewfish instead.

**Comment by redonkulus** [comment]
:+1: maybe add some comments on the internal structure changes.

**Pull request merged**